src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -0,0 +1,158 @@
+    status = ucs_socket_connect(cep->fd, server_addr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        goto err_close_socket;
+    }
+    ucs_assert((status == UCS_OK) || (status == UCS_INPROGRESS));
+
+    if (status == UCS_OK) {
+        cep->state |= UCT_TCP_SOCKCM_EP_CONNECTED;
+        /* TODO: start sending the user's data */
+
+        goto out_print;
+    }
+
+    sa_arg_ctx = ucs_malloc(sizeof(uct_tcp_sa_arg_t), "client sa_arg_ctx");
+    if (sa_arg_ctx == NULL) {
+        ucs_error("failed to allocate memory for a client_ctx");
+        status = UCS_ERR_NO_MEMORY;
+        goto err_close_socket;
+    }
+
+    sa_arg_ctx->fd = cep->fd;
+    sa_arg_ctx->ep = cep;
+
+    /* Adding the arg to a list on the cm for cleanup purposes */
+    ucs_list_add_tail(&tcp_sockcm->sa_arg_list, &sa_arg_ctx->list);
+
+    async_ctx = tcp_sockcm->super.iface.worker->async;
+    status    = ucs_async_set_event_handler(async_ctx->mode, cep->fd,
+                                            /* wait until connect() completes */
+                                            UCS_EVENT_SET_EVWRITE,
+                                            uct_tcp_sa_data_handler, sa_arg_ctx,
+                                            async_ctx);
+    if (status != UCS_OK) {
+        goto err_free_sa_arg_ctx;
+    }
+
+out_print:
+    ucs_debug("created a TCP SOCKCM endpoint (fd=%d) on tcp cm %p, "
+              "remote addr: %s", cep->fd, tcp_sockcm,
+              ucs_sockaddr_str(server_addr, ip_port_str, UCS_SOCKADDR_STRING_LEN));
+
+    return status;
+
+err_free_sa_arg_ctx:
+    ucs_list_del(&sa_arg_ctx->list);
+    free(sa_arg_ctx);
+err_close_socket:
+    close(cep->fd);
+err:
+    return status;
+}
+
+UCS_CLASS_INIT_FUNC(uct_tcp_sockcm_ep_t, const uct_ep_params_t *params)
+{
+    ucs_status_t status;
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_cm_base_ep_t, params);
+
+    self->state           = 0;
+    self->comm_ctx.buf    = NULL;
+    self->comm_ctx.offset = 0;
+
+    if (params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR) {
+        status = uct_tcp_sockcm_ep_client_init(self, params);
+    } else if (params->field_mask & UCT_EP_PARAM_FIELD_CONN_REQUEST) {
+        status = uct_tcp_sockcm_ep_server_init(self, params);
+    } else {
+        ucs_error("either UCT_EP_PARAM_FIELD_SOCKADDR or UCT_EP_PARAM_FIELD_CONN_REQUEST "
+                  "has to be provided");
+        status = UCS_ERR_INVALID_PARAM;
+    }
+
+    if (status == UCS_OK) {
+        ucs_debug("created an endpoint on tcp_sockcm %p id: %d",
+                  uct_tcp_sockcm_ep_get_cm(self), self->fd);
+    }
+
+    return status;
+}
+
+UCS_CLASS_CLEANUP_FUNC(uct_tcp_sockcm_ep_t)
+{
+    uct_tcp_sockcm_t *tcp_sockcm = uct_tcp_sockcm_ep_get_cm(self);
+    ucs_status_t status;
+
+    UCS_ASYNC_BLOCK(tcp_sockcm->super.iface.worker->async);
+
+    status = ucs_async_remove_handler(self->fd, 1);
+    if (status != UCS_OK) {
+        ucs_debug("failed to remove event handler for fd %d: %s",
```

Comment:
redundant, ucs_async_remove_handler already prints a warning

Reply:
coverity complains if the return value isn't checked from this function in some places but is checked in others.